File: java/239_SlidingWindowMaximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        if (nums.length == 0 || nums == null || k == 0) {
            return new int[0];
        }
        // 递减单调队
        Deque<Integer> deque = new LinkedList<>();
        int[] res = new int[nums.length - k + 1];
        for (int i = 1 - k, j = 0; j < nums.length; i++, j++) {
            // 最大值在左边
            if (i > 0 && deque.peekFirst() == nums[i-1]) {
                deque.removeFirst();
            }
            // 保持递减
            while (!deque.isEmpty() && deque.peekLast() < nums[j]) {
                deque.removeLast();
            }
            deque.addLast(nums[j]);
            if (i >= 0) {
                res[i] = deque.peekFirst();
            }
        }
        return res;
    }
}

File: c++/239_SlidingWindowMaximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        deque<int> dq;
        vector<int> res;
        for (int i = 0; i < nums.size(); i++) {
            if (dq.size() && i-k+1 > dq.front()) dq.pop_front();
            while (dq.size() && nums[i] >= nums[dq.back()]) dq.pop_back();
            dq.push_back(i);
            if (i >= k-1) res.push_back(nums[dq.front()]);
        }
        return res;
    }
};
